File: src/lib/api/providerCenter.ts
import { invoke } from "@tauri-apps/api/core";
import type { Provider } from "@/types";

export type ProviderCenterApp =
  | "claude"
  | "claude-desktop"
  | "codex"
  | "gemini"
  | "grokbuild"
  | "opencode"
  | "openclaw"
  | "hermes"
  | "pi";

export interface ProviderDefinition {
  id: string;
  name: string;
  protocol: string;
  baseUrl: string;
  models: string[];
  discoveredModels: string[];
  notes: string;
  enabled: boolean;
  revision: number;
  source?: {
    sourceApp: string;
    sourceRef: string;
    importedAt: number;
    sourceFingerprint?: string;
    lastObservedAt?: number;
  };
  sources: Array<{
    sourceApp: string;
    sourceRef: string;
    importedAt: number;
    sourceFingerprint?: string;
    lastObservedAt?: number;
  }>;
  credentialConfigured: boolean;
  credentialHint?: string;
  lastDiscoveryAt?: number;
  lastDiscoveryError?: string;
  createdAt: number;
  updatedAt: number;
}

export interface ProviderBinding {
  providerId: string;
  appType: string;
  status:
    | "pending"
    | "applied"
    | "overridden"
    | "unsupported"
    | "drifted"
    | "detached"
    | "failed";
  enabled: boolean;
  overrideEnabled: boolean;
  appliedRevision?: number;
  expectedFingerprint?: string;
  lastError?: string;
  lastTransactionId?: string;
  updatedAt: number;
}

export interface ProviderCenterState {
  definitions: ProviderDefinition[];
  bindings: ProviderBinding[];
  transactions: ProviderApplyTransaction[];
}

export type AgentProviderScope = "universal" | "agentOnly" | "nativeAccount";

export type AgentProviderOwnership =
  | "providerCenterProjection"
  | "ccSwitchManaged"
  | "agentNative";

export interface AgentProviderCatalogItem {
  providerId: string;
  providerName: string;
  category?: string;
  scope: AgentProviderScope;
  ownership: AgentProviderOwnership;
  definitionId?: string;
  bindingStatus?: ProviderBinding["status"];
  appliedRevision?: number;
  drifted: boolean;
  readOnly: boolean;
}

export interface AgentProviderCatalog {
  appType: ProviderCenterApp;
  items: AgentProviderCatalogItem[];
  generatedAt: number;
}

export interface ImportConflict {
  existingProviderId: string;
  existingName: string;
  existingRevision: number;
  reasons: string[];
}

export type ImportCommitAction = "createCopy" | "merge" | "skip";

export interface ImportCommitDecision {
  action: ImportCommitAction;
  targetProviderId?: string;
  expectedRevision?: number;
}

export interface ImportCommitResult {
  action: ImportCommitAction;
  provider?: ProviderDefinition;
  repeated: boolean;
}

export interface ImportCandidate {
  id: string;
  sessionId: string;
  sourceRef: string;
  sourceApp: string;
  sourceKind: "externalManaged" | "providerCenterProjection" | "localManaged";
  name: string;
  protocol: string;
  baseUrl: string;
  models: string[];
  credentialConfigured: boolean;
  credentialHint?: string;
  conflicts: ImportConflict[];
}

export interface ImportFailure {
  appType: string;
  sourceRef?: string;
  code: string;
  stage: string;
  message: string;
}

export interface ProviderImportSession {
  id: string;
  state: "ready" | "readyWithErrors" | "completed" | "expired" | string;
  candidates: ImportCandidate[];
  errors: ImportFailure[];
  createdAt: number;
  expiresAt: number;
}

export interface SaveProviderDefinitionInput {
  id?: string;
  name: string;
  protocol: string;
  baseUrl: string;
  models: string[];
  notes?: string;
  enabled?: boolean;
  expectedRevision?: number;
  credentialAction?: "keep" | "replace" | "clear";
  /** Write-only. It is never returned by any Provider Center API. */
  apiKey?: string;
  appTypes: string[];
}

export interface ModelDiscoveryResult {
  providerId: string;
  models: string[];
  discoveredAt: number;
  error?: string;
}

export interface UnifiedModelCatalogEntry {
  id: string;
  appType: string;
  providerId: string;
  providerName: string;
  modelId: string;
  protocol: string;
  sourceType: "apiKey" | "nativeAccount";
  isDefault: boolean;
}

export interface UnifiedModelCatalog {
  entries: UnifiedModelCatalogEntry[];
  generatedAt: number;
}

export interface ProviderApplyPreviewTarget {
  appType: string;
  operation: "create" | "update" | "unsupported";
  compatible: boolean;
  connectionMode: "direct" | "proxy" | "unsupported";
  routeId?: string;
  requiresTakeover: boolean;
  drifted: boolean;
  currentProviderId?: string;
  liveFingerprint?: string;
  message?: string;
}

export interface ProviderApplyPreview {
  token: string;
  providerId: string;
  providerRevision: number;
  targets: ProviderApplyPreviewTarget[];
  createdAt: number;
}

export interface ProviderApplyTargetResult {
  appType: string;
  status: string;
  message?: string;
}

export interface ProviderApplyTransaction {
  id: string;
  providerId: string;
  providerRevision: number;
  status: string;
  targets: ProviderApplyTargetResult[];
  createdAt: number;
  completedAt?: number;
}

export interface ManagedProviderDraftInput {
  appType: string;
  provider: Provider;
  definitionId: string;
  expectedRevision?: number;
  /** Agents to save the universal provider to. If empty, defaults to [appType]. */
  targetAppTypes?: string[];
}

export type DeleteMode = "removeCurrent" | "detachKeepIndependent" | "deleteGlobally";

export const providerCenterApi = {
  get: (): Promise<ProviderCenterState> => invoke("get_provider_center"),
  getAgentProviderCatalog: (
    appType: ProviderCenterApp,
  ): Promise<AgentProviderCatalog> =>
    invoke("get_provider_center_agent_provider_catalog", { appType }),
  getModelCatalog: (appTypes: string[] = []): Promise<UnifiedModelCatalog> =>
    invoke("get_provider_center_model_catalog", { appTypes }),
  save: (input: SaveProviderDefinitionInput): Promise<ProviderDefinition> =>
    invoke("save_provider_center_definition", { input }),
  delete: (providerId: string): Promise<void> =>
    invoke("delete_provider_center_definition", { providerId }),
  deleteProvider: (
    providerId: string,
    appType: string,
    mode: DeleteMode,
  ): Promise<void> =>
    invoke("delete_provider_center_provider", { providerId, appType, mode }),
  duplicate: (providerId: string): Promise<ProviderDefinition> =>
    invoke("duplicate_provider_center_definition", { providerId }),
  discoverModels: (providerId: string): Promise<ModelDiscoveryResult> =>
    invoke("discover_provider_center_models", { providerId }),
  scanImports: (): Promise<ImportCandidate[]> =>
    invoke("scan_provider_center_imports"),
  startImportSession: (
    appTypes: string[] = [],
  ): Promise<ProviderImportSession> =>
    invoke("start_provider_center_import_session", { appTypes }),
  getImportSession: (sessionId: string): Promise<ProviderImportSession> =>
    invoke("get_provider_center_import_session", { sessionId }),
  commitImportCandidate: (
    sessionId: string,
    candidateId: string,
    appTypes: string[],
    decision: ImportCommitDecision,
  ): Promise<ImportCommitResult> =>
    invoke("commit_provider_center_import_candidate", {
      sessionId,
      candidateId,
      appTypes,
      decision,
    }),
  importCandidate: (
    sourceRef: string,
    appTypes: string[],
  ): Promise<ProviderDefinition> =>
    invoke("import_provider_center_candidate", { sourceRef, appTypes }),
  attach: (
    providerId: string,
    appType: ProviderCenterApp,
  ): Promise<ProviderBinding> =>
    invoke("attach_provider_center_binding", { providerId, appType }),
  apply: (providerId: string, appTypes: string[]): Promise<ProviderBinding[]> =>
    invoke("apply_provider_center_bindings", { providerId, appTypes }),
  previewApply: (
    providerId: string,
    appTypes: string[],
  ): Promise<ProviderApplyPreview> =>
    invoke("preview_provider_center_apply", { providerId, appTypes }),
  previewDefinitionCompatibility: (
    providerId: string,
    appTypes: string[],
  ): Promise<ProviderApplyPreview> =>
    invoke("preview_provider_center_definition_compatibility", {
      providerId,
      appTypes,
    }),
  applyTransaction: (
    providerId: string,
    appTypes: string[],
    previewToken: string,
    idempotencyKey = crypto.randomUUID(),
  ): Promise<ProviderApplyTransaction> =>
    invoke("apply_provider_center_transaction", {
      providerId,
      appTypes,
      previewToken,
      idempotencyKey,
    }),
  restoreTransaction: (
    transactionId: string,
  ): Promise<ProviderApplyTransaction> =>
    invoke("restore_provider_center_transaction", { transactionId }),
  setOverride: (
    providerId: string,
    appType: string,
    enabled: boolean,
  ): Promise<void> =>
    invoke("set_provider_center_binding_override", {
      providerId,
      appType,
      enabled,
    }),
  disableBinding: (
    providerId: string,
    appType: string,
    removeProjection = false,
  ): Promise<void> =>
    invoke("disable_provider_center_binding", {
      providerId,
      appType,
      removeProjection,
    }),
  previewManagedDraft: (
    input: ManagedProviderDraftInput,
  ): Promise<ProviderApplyPreview> =>
    invoke("preview_provider_center_managed_draft", { input }),
  confirmManagedDraft: (
    input: ManagedProviderDraftInput,
    previewToken: string,
    idempotencyKey = crypto.randomUUID(),
  ): Promise<ProviderApplyTransaction> =>
    invoke("confirm_provider_center_managed_draft", {
      input,
      previewToken,
      idempotencyKey,
    }),
};

File: src-tauri/src/commands/provider_center.rs
use crate::provider_center::{
    self, ImportCandidate, ImportCommitDecision, ImportCommitResult, ModelDiscoveryResult,
    ProviderApplyPreview, ProviderApplyTransaction, ProviderBinding, ProviderCenterOperationState,
    ProviderCenterState, ProviderDefinition, ProviderImportSession, SaveProviderDefinitionInput,
    UnifiedModelCatalog,
};
use crate::store::AppState;
use tauri::State;

#[tauri::command]
pub async fn get_provider_center(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
) -> Result<ProviderCenterState, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::state(state.inner()).map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_provider_center_model_catalog(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] appTypes: Option<Vec<String>>,
) -> Result<UnifiedModelCatalog, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::unified_model_catalog(state.inner(), appTypes.unwrap_or_default())
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn save_provider_center_definition(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    input: SaveProviderDefinitionInput,
) -> Result<ProviderDefinition, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::save_definition(state.inner(), input).map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn delete_provider_center_definition(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
) -> Result<(), String> {
    let _data_guard = operations.lock_data().await;
    provider_center::delete_definition(state.inner(), &providerId)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn duplicate_provider_center_definition(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
) -> Result<ProviderDefinition, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::duplicate_definition(state.inner(), &providerId)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn discover_provider_center_models(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
) -> Result<ModelDiscoveryResult, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::discover_models(state.inner(), &providerId)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn scan_provider_center_imports(
    state: State<'_, AppState>,
) -> Result<Vec<ImportCandidate>, String> {
    provider_center::scan_imports(state.inner()).map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn start_provider_center_import_session(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] appTypes: Option<Vec<String>>,
) -> Result<ProviderImportSession, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::start_import_session(state.inner(), appTypes.unwrap_or_default())
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_provider_center_import_session(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] sessionId: String,
) -> Result<ProviderImportSession, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::get_import_session(state.inner(), &sessionId)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn commit_provider_center_import_candidate(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] sessionId: String,
    #[allow(non_snake_case)] candidateId: String,
    #[allow(non_snake_case)] appTypes: Vec<String>,
    decision: ImportCommitDecision,
) -> Result<ImportCommitResult, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::commit_import_session_candidate(
        state.inner(),
        &sessionId,
        &candidateId,
        appTypes,
        decision,
    )
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn import_provider_center_candidate(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] sourceRef: String,
    #[allow(non_snake_case)] appTypes: Vec<String>,
) -> Result<ProviderDefinition, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::import_candidate(state.inner(), &sourceRef, appTypes)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn apply_provider_center_bindings(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] appTypes: Vec<String>,
) -> Result<Vec<ProviderBinding>, String> {
    let _data_guard = operations.lock_data().await;
    let targets =
        provider_center::apply_target_app_types(state.inner(), &providerId, appTypes.clone())
            .map_err(|error| error.to_string())?;
    let _guards = operations.lock_apps(targets).await;
    provider_center::apply_bindings(state.inner(), &providerId, appTypes)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn preview_provider_center_apply(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] appTypes: Vec<String>,
) -> Result<ProviderApplyPreview, String> {
    let _data_guard = operations.lock_data().await;
    provider_center::preview_apply(state.inner(), &providerId, appTypes)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn apply_provider_center_transaction(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] appTypes: Vec<String>,
    #[allow(non_snake_case)] previewToken: String,
    #[allow(non_snake_case)] idempotencyKey: Option<String>,
) -> Result<ProviderApplyTransaction, String> {
    let _data_guard = operations.lock_data().await;
    let targets =
        provider_center::apply_target_app_types(state.inner(), &providerId, appTypes.clone())
            .map_err(|error| error.to_string())?;
    let _guards = operations.lock_apps(targets).await;
    provider_center::apply_transaction(
        state.inner(),
        &providerId,
        appTypes,
        &previewToken,
        idempotencyKey.as_deref(),
    )
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn restore_provider_center_transaction(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] transactionId: String,
) -> Result<ProviderApplyTransaction, String> {
    let _data_guard = operations.lock_data().await;
    let targets = provider_center::restore_target_app_types(state.inner(), &transactionId)
        .map_err(|error| error.to_string())?;
    let _guards = operations.lock_apps(targets).await;
    provider_center::restore_transaction(state.inner(), &transactionId)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn set_provider_center_binding_override(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] appType: String,
    enabled: bool,
) -> Result<(), String> {
    let _data_guard = operations.lock_data().await;
    provider_center::set_binding_override(state.inner(), &providerId, &appType, enabled)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn disable_provider_center_binding(
    state: State<'_, AppState>,
    operations: State<'_, ProviderCenterOperationState>,
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] appType: String,
    #[allow(non_snake_case)] removeProjection: bool,
) -> Result<(), String> {
    let _data_guard = operations.lock_data().await;
    let _guards = operations.lock_apps(vec![appType.clone()]).await;
    provider_center::disable_binding(state.inner(), &providerId, &appType, removeProjection)
        .map_err(|error| error.to_string())
}
